custom/src/main/java/com/splunk/opentelemetry/servicename/AppServerServiceNameDetector.java

```diff
@@ -0,0 +1,262 @@
+
+  private String handleExploded(Path path, Path descriptor, DescriptorHandler handler) {
+    if (Files.isRegularFile(descriptor)) {
+      return handle(() -> Files.newInputStream(descriptor), path, handler);
+    }
+
+    return null;
+  }
+
+  private String handle(InputStreamSupplier supplier, Path path, DescriptorHandler handler) {
+    try {
+      try (InputStream inputStream = supplier.supply()) {
+        String candidate = parseDescriptor(inputStream, handler);
+        if (isValidResult(path, candidate)) {
+          return candidate;
+        }
+      }
+    } catch (Exception exception) {
+      log.warn("Failed to parse descriptor", exception);
+    }
+
+    return null;
+  }
+
+  private static String parseDescriptor(InputStream inputStream, DescriptorHandler handler)
+      throws ParserConfigurationException, SAXException, IOException {
+    if (SaxParserFactoryHolder.saxParserFactory == null) {
+      return null;
+    }
+    SAXParser saxParser = SaxParserFactoryHolder.saxParserFactory.newSAXParser();
+    saxParser.parse(inputStream, handler);
+    return handler.displayName;
+  }
+
+  private interface InputStreamSupplier {
+    InputStream supply() throws IOException;
+  }
+
+  private static class WebXmlHandler extends DescriptorHandler {
+
+    WebXmlHandler() {
+      super("web-app");
+    }
+  }
+
+  private static class ApplicationXmlHandler extends DescriptorHandler {
+
+    ApplicationXmlHandler() {
+      super("application");
+    }
+  }
+
+  private static class DescriptorHandler extends DefaultHandler {
+    private final String rootElementName;
+    private final Deque<String> currentElement = new ArrayDeque<>();
+    private boolean setDisplayName;
+    String displayName;
+
+    DescriptorHandler(String rootElementName) {
+      this.rootElementName = rootElementName;
+    }
+
+    @Override
+    public void startElement(String uri, String localName, String qName, Attributes attributes) {
+      if (displayName == null
+          && rootElementName.equals(currentElement.peek())
+          && "display-name".equals(qName)) {
+        String lang = attributes.getValue("xml:lang");
+        if (lang == null || "".equals(lang)) {
+          lang = "en"; // en is the default language
+        }
+        if ("en".equals(lang)) {
+          setDisplayName = true;
+        }
+      }
+      currentElement.push(qName);
+    }
+
+    @Override
+    public void endElement(String uri, String localName, String qName) {
+      currentElement.pop();
+      setDisplayName = false;
+    }
+
+    @Override
+    public void characters(char[] ch, int start, int length) {
+      if (setDisplayName) {
+        displayName = new String(ch, start, length);
+      }
+    }
+  }
+
+  private static class SaxParserFactoryHolder {
+    private static final SAXParserFactory saxParserFactory = getSaxParserFactory();
+
+    private static SAXParserFactory getSaxParserFactory() {
+      try {
+        return SAXParserFactory.newInstance();
```

Comment:
Odd implementation -- I'd have expected a holder to hold and return a single instance rather than instantiating new ones each time.

Reply:
Oh, I see it now, it's just direct referenced. 🤷🏻